### ISSUE TITLE
Designing feature combo options

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#10624]
I've been thinking about how to represent feature combos for a while, since it comes up from time to time, and today ran into a case where it's what we really want. the cmake I'm porting does something like this:
```cmake
set(PROJECT_OPTION_AUTO TRUE CACHE BOOL "autodect which dependency to use")
set(PROJECT_OPTION_DEP_1 FALSE CACHE BOOL "Use dep 1 for feature foo")
set(PROJECT_OPTOIN_DEP_2 FALSE CACHE BOOL "Use dep 2 for feature foo")
set(PROJECT_OPTION_DEP_3 FALSE CACHE BOOL "use dep 3 for feature foo")
```

You can write the same code in meson, but you end up doing a lot of code to validate that only one is enabled(), or that not all of them are disabled(). So I was thinking about it and came up with this:
```meson
option(
  'foo_dep',
  type : 'combo_feature',
  choices : ['1', '2', '3'],
)
```
When you set this the command line looks like this (the option takes the form `{name.choice_value}`):
```sh
meson setup builddir -Dfoo_dep.2=enabled`
```

When any option is set to enabled, then the others are automatically disabled. So this is equivalent to setting `-Dfoo_dep1=disabled -Dfoo_dep.3=disabled -Dfoo_dep.1=enabled`

```
meson setup builddir -Dfoo_dep.2=disabled -Dfoo_dep.1=auto`
```
However, specific options may be disabled leaving others as auto. This allows you to exclude dep2 if you think it's an inferior solution, but still allow dep1 and dep3 to be tried.

Since `.` is a reserved value in meson options, we have no chance of a name collsion from the auto-generated options names.

The only question I have is how to handle the default value. I can think of two ways, and both have drawbacks:
1. `value : '2'`: means that 2 is enabled by default and `1` and `3` are disabled by default
2. `value : ['auto', 'disabled', 'auto']`: each option is given a default value in an array, coresponding to the option. more flexible, but more verbose, also requires validation that two options aren't enabled at the same time

The final thought I had is a `required` field which would make it an error to disable all of the options, though thats a little niche and could probably be handled in the meson.build easily enough with something like:
```meson
if not (dep1.found() or dep2.found() or dep3.found())
  error(...)
endif
```

Before I go writing a bunch of code, does anyone have objections, questions, or better ideas?

**Comment by tristan957** [comment]
It would be nice if you provide how one would currently do this in Meson. Not really visualizing it all that well.

**Comment by jpakkane** [comment]
Having two different combo options for doing roughly the same is confusing. Would it be possible to achieve the same with some other primitives? Like having some way to get the combo selection as a `[false, false, true, false false]` array or something?

**Comment by dcbaker** [comment]
> Having two different combo options for doing roughly the same is confusing.

I'm not sure I understand what you mean. A feature combo really isn't any more confusing than feature vs combo itself. You can achieve basically the same thing with a combo that you can with a feature, it just takes a lot more code. A combo of features can be achieved with a series of combos, just with a lot more code.

>  Like having some way to get the combo selection as a [false, false, true, false false] array or something?

No, that wouldn't help at all. I'm dealing with mutually exclusive dependencies that can all provide the same functionality, features are what I want, but without having to write a ton of biolerplate to ensure that the state is correct. The case I have in mind is this:

meson_options.txt:
```meson
option(
  'dep1',
  type : 'feature',
  description : 'use dep 1 to provide foo feature',
)
option(
  'dep2',
  type : 'feature',
  description : 'use dep 2 to provide foo feature',
)
option(
  'dep3',
  type : 'feature',
  description : 'use dep 3 to provide foo feature',
)
```

meson.build:
```meson
cc = meson.get_compiler('c')
dep1 = get_option('dep1')
dep2 = get_option('dep2')
dep3 = get_option('dep3')
if dep1.disabled() and dep2.disabled() and dep3.disabled()
  error('at least one of dep1, dep2, and dep3 must be allowed')
endif
# because these are exclusive, you cannot "enable" (thus require) more than one
if dep1.enabled().to_int() + dep2.enabled().to_int() + dep3.enabled().to_int() > 1
  error('only one of dep1, dep2, and dep3 may be enabled at a time')
endif
foo_dep = dependency('dep1', required : dep1)
if not foo_dep.found()
  foo_dep = dependency('dep2', required : dep2)
  if not foo_dep.found()
    foo_dep = cc.find_library('dep3', required : true)
  endif
endif
if not foo_dep.found()
  error('must have dep1, dep2, or dep3')
endif
```

Which we could replace with;
meson_options.txt
```meson
option(
  'foo_dep',
  type : 'feature_combo',
  description : 'which provider for foo to use',
  required : true
)
```
meson.build:
```meson
cc = meson.get_compiler('c')
foo_dep = dependency('dep1', required : get_option('foo_dep.1')
if not foo_dep.found()
  foo_dep = dependency('dep2', required : get_option('foo_dep.2')
  if not foo_dep.found()
    foo_dep = cc.find_library('dep3', required : get_option('foo_dep.3')
  endif
endif
```

**Comment by jpakkane** [comment]
Isn't the last the same as:

```meson
dep_name = get_option('foo_dep')
foo_dep = dependency('dep1', required : dep_name == 'name1')
if not foo_dep.found()
  foo_dep = dependency('dep2', required : dep_name == 'name2')
  if not foo_dep.found()
    foo_dep = cc.find_library('dep3', required : dep_name == 'name3')
  endif
endif
if not foo_dep.found()
  error('must have name1, name2, or name3')
endif
```

Where `meson_options.txt` has a single combo with possible values `name1', 'name2' and 'name3'.

**Comment by dcbaker** [comment]
No, the difference is that only one option is allowed to be enabled() and the others are forced to disabled(), but any number could be auto(). so you could pass the following options and get the following results:
`-Dfoo_dep.1=enabled` == `dependency('dep1', required : true)`  # this implies `-Dfoo_dep.2=disabled -Dfoo_dep.3=disabled`
`-Dfoo_dep.2=disabled` == `dependency('dep1', 'dep3')`  # .1 and .3 remain auto().
`-Dfoo_dep.1=disabled -Dfoo_dep.2=disabled` = `dependency('dep3')` # .3 remains auto()\

Basically this allows you to provide for "any one of these will work, but we must have one, and the use needs to be able to control which one(s) are allowed. In your example there's no way for the end user to say "never use dep1", only "you must use dep1, or you may use dep1", but if dep1 is found dep2 will never be tried, even if `dep_name == 'name2'`

**Comment by jpakkane** [comment]
Can you post the details on what the original practical requirement is? That is with correct dependency names and all that?

**Comment by dcbaker** [comment]
https://github.com/bombela/backward-cpp/blob/master/BackwardConfig.cmake is the cmake I'm looking at. To port this to meson I came up with a (somewhat more complicated) version of the example code
